framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -107,75 +117,117 @@ public ValueTransfer(CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis)
+            if (range.isWiderThan(ValueAnnotatedTypeFactory.MAX_VALUES)) {
+                return new ArrayList<Character>();
+            } else {
+                return ValueCheckerUtils.getValuesFromRange(range, Character.class);
+            }
+        }
+
+        return new ArrayList<Character>();
     }
 
+    /**
+     * Get possible numerical values from annotation and cast them to the underlying type. Return
+     * null if @BottomVal.
+     */
     private List<? extends Number> getNumericalValues(
             Node subNode, TransferInput<CFValue, CFStore> p) {
         CFValue value = p.getValueOfSubNode(subNode);
-        AnnotationMirror numberAnno =
-                AnnotationUtils.getAnnotationByClass(value.getAnnotations(), IntVal.class);
-        List<? extends Number> values;
-        if (numberAnno == null) {
-            numberAnno =
-                    AnnotationUtils.getAnnotationByClass(value.getAnnotations(), DoubleVal.class);
-            if (numberAnno != null) {
-                values =
-                        AnnotationUtils.getElementValueArray(
-                                numberAnno, "value", Double.class, true);
-            } else {
-                return new ArrayList<Number>();
-            }
-        } else {
-            values = AnnotationUtils.getElementValueArray(numberAnno, "value", Long.class, true);
+        AnnotationMirror numberAnno;
+
+        numberAnno = AnnotationUtils.getAnnotationByClass(value.getAnnotations(), IntVal.class);
+        if (numberAnno != null) {
+            List<Long> values =
+                    AnnotationUtils.getElementValueArray(numberAnno, "value", Long.class, true);
+            return NumberUtils.castNumbers(subNode.getType(), values);
         }
-        return NumberUtils.castNumbers(subNode.getType(), values);
-    }
 
-    private AnnotationMirror createStringValAnnotationMirror(List<String> values) {
-        if (values.isEmpty()) {
-            return ((ValueAnnotatedTypeFactory) atypefactory).UNKNOWNVAL;
+        numberAnno = AnnotationUtils.getAnnotationByClass(value.getAnnotations(), DoubleVal.class);
+        if (numberAnno != null) {
+            List<Double> values =
+                    AnnotationUtils.getElementValueArray(numberAnno, "value", Double.class, true);
+            return NumberUtils.castNumbers(subNode.getType(), values);
         }
-        return ((ValueAnnotatedTypeFactory) atypefactory).createStringAnnotation(values);
+
+        numberAnno = AnnotationUtils.getAnnotationByClass(value.getAnnotations(), BottomVal.class);
+        if (numberAnno != null) {
+            return null;
+        }
+
+        return new ArrayList<Number>();
     }
 
-    private AnnotationMirror createNumberAnnotationMirror(List<Number> values) {
-        if (values.isEmpty()) {
-            return ((ValueAnnotatedTypeFactory) atypefactory).UNKNOWNVAL;
+    private Range getIntRange(Node subNode, TransferInput<CFValue, CFStore> p) {
```

Comment:
Please add a Javadoc comment for every method that you added or modified.
That will make the code easier to read.